### ISSUE TITLE
Deterioration - a git issue for the sole purpose of discussing it

### ISSUE THREAD

**Issue by NilanthAnimosus** [Regalis11/Barotrauma#1361]
**Description**
First and foremost - we need feedback from most players, alpha or otherwise on the deterioration system and would like to see this feedback as suggestions or thoughts in this git issue. The mechanic as is:
- A time-to-fix duration timer is set on a device, one for high skill and one for low
- The timer is bugged in the latest steam version, IE. everything is 5 seconds or 100 seconds for high or low skill.
- A junction will take 5 seconds to fix with skill, 10 seconds without the skill due to the fixes for the next version (post 0.8.9.7 steam closed alpha).
- Said junction will take between 120 seconds to 1000 seconds (2 minutes to 16.66 minutes) before it begins to deteriorate. yes, these are 0.8.9.7 numbers on what it should be for a stock unedited junction in a new submarine.
- The default junction deterioration speed is 0.1f, which translates to 0.1% condition lost per second, or apparently 16.6 minutes to break, this doesn't feel right though but this is how it is set.

After a user created a very long document on this, parts of which I did and did not agree with or see different ways to potentially handle and not make the game suddenly easier or always doable, but also not as much of a grind or boring issue. I decided I'll put up a git issue where everyone can toss in feedback, make suggestions, state thoughts and such for the deterioration mechanic.

I'm going to include the document written by Ol stinky: [Barotrauma_Deterioration_Suggestions.docx](https://github.com/Regalis11/Barotrauma/files/3012697/Barotrauma_Deterioration_Suggestions.docx)

**My thoughts on these issues**
Personally, I agree that deterioration is easily too punishing for small crews, but I also disagree on it taking too long when the last engineer/mechanic dies as this would normally (and should be more normally) a game over condition in legacy, and I feel if its too much to repair then the game probably should have ended but players never know when to quit, so it feels too much or too boring. - giving us the option to fix everything instead makes everyone watch those 100 second timers (10-20 seconds on no skill with the fixes to timers, which if you ask me makes deterioration more pointless and removes why we even have mechanics and engineers).

I do believe deterioration should factor in a devices usage - I don't see why the deterioration should start for a device never used every 2 minutes to 10 minutes. I also don't see why out of 12 junctions 3 will begin decaying, then in 1 minute 2 more, then in 2 minutes another 3, and then in half a minute two I already fixed and 2 other ones. - this does cause the mentioned problem of punishing small crews and consuming more time than it ever needed to.  I also dont feel deterioration should be as cookie cut as 100% -> 0% loss in one go with the only chance to notice it being the decrease in between.

What do I think could be a good way to stop this and improve the system personally?

***What I think could be done to alleviate this somewhat in the form of suggestions***

most of these are in an order and more intended to potentially combine into a new deterioration repair system. one I think may be better than the current if it encompassed this as a whole, the new complexity wouldn't necessarily be a burden on new players either I believe.

- [ ] **Have deterioration work in stages, so the first deterioration lowers it by 20% and initiates another timer, say in 20% stages), have the first repair available at the first step lost (80%), the GUI should also be tweaked slightly to represent these zones and where decay will begin to become an issue.**

Why do this? it may seem artificial but its a lot less painful this way, devices going from 100% condition to 0% condition and at random is so much more of a pain than devices losing a little condition periodically to no ill effect, I would rather find 5 devices needing 20% and 2 devices needing 40%, than 5 devices I just may have repaired in the same room I just finished already at or on their way to 0%. its less maintenance than sub-rot.

- [ ] **Base deterioration speed on the usage of the component where valid.**

A powered down submarine or out of use fabricator shouldn't be deteriorating much at all. a submarine not moving or pumping shouldn't be breaking its engines and pumps. a short swim to the ruins is not 2 years of decay and we do not have metal eating space termites on our submarine.

If a fabricator is in use, or junctions are powered on - this should deteriorate it faster, if a device is powered down or not in use, it should deteriorate slower. ammo loaders actively used should take hits to their timers and autopilot should cost more deterioration than manual, and maintain should take the least.

A moving submarine should more rapidly expend its fix cool-downs than one not utilizing those engines, eternally on pumps should take more maintenance too. this means more maintenance when in use and traveling, and less maintenance when ruins exploring or maintaining. maintenance would also come after the fights, not during at all times due to mechanisms like reductions on timers for having used the ammo loaders.

- [ ] **Scale the repair timers to the amount of skill even if insufficient (Like crafting), if using repair steps additionally scale this timer to the step your repairing if skill is insufficient, have repair timers be based on the amount you wish to repair (Less filled = faster repaired) to allow for 'hot fixing' past one of the repair steps.**

Some examples of this:
5 seconds (10 at most) to fully repair with the skill necessary for all steps (1-2s / step as example which lasts numerous minutes per step)

with half the skill:
10 seconds to repair to the first step without skill (20% condition - give a full cool down timer if it decays to this step)
20 seconds to repair the second step (40% condition - total 30 seconds)
30 seconds for the third step (60% condition, part works flawlessly is using the system below - total 1 minute)
40 seconds for the fourth step (80% condition, total 1 minute 40 seconds)
50 seconds for the fifth step (100% condition, total 2 minutes 30 seconds).

with three quarters or more of the skill:
5 seconds to repair to the first step without skill (20% condition - give a full cool down timer if it decays to this step)
10 seconds to repair the second step (40% condition - total 15 seconds)
15 seconds for the third step (60% condition, part works flawlessly is using the system below - total 30 seconds)
20 seconds for the fourth step (80% condition, total 50 seconds)
25 seconds for the fifth step (100% condition, total 1 minute 15 seconds).

It is totally unnecessary to repair past the third step for none-classes, it barely matters to repair past the second step. and its perfectly functional with cool down at the first step. the above timers should be very long for a full fix and demand the related job for peak repairs, and the skillgain scaled to how long this repair is (Less for far longer repairs so its more of a waste) to make it less of a reason to actually do. in legacy we could never repair these devices, I don't see why every job should be able to repair something to full in 20 seconds for the next 5-10 minutes. but I do see why every job should be able to repair everything to a minimum effectiveness. its not an issue of reducing repair timers or making everything easier, but changing where one needs to reach or how much to do, and making this obvious.

- [ ] **Give all repairs a small minimum cool down timer it cannot deteriorate for on top of the standard timer, IE. repair 53% - does not need repairs for 1 minute, give people time to fix devices and make unskilled fixing worth something but still need a related job for it later.**

if using steps, this still means a minute cooldown before deteriorating from halfway between steps to the one you passed and hitting that steps timer instead, IE steeling from 53% to 40% after 1 minute + some decay. this as a side effect means if each step took 2-5 minutes, hitting 2/5 means 1 minute minimum, then decaying to each of the + 2-5 minutes per step (so a 50% repair is about 5-11 minutes to break fully, if using the scaled repairs, this is only 30 seconds to have patched up).

- [ ] **Modify the effectiveness of equipment based on its condition beyond a certain point (IE. below 50%, this should take multiple missed repairs).**

This goes hand in hand with the concept of lock-stepped repairs. examples by component type including where mods could add decay values to not be exclusive:

1.) Fabricator/deconstructors timers increased by condition damage.

2.) Batteries gain less charge and have reduced maximum output

3.) Pumps have reduced drain/pump weights.

4.) Power transfer (IE. junctions) output a reduced voltage for condition damage to connected components but not other power transfer (Same power, reduced voltage to connected devices that are not power transfer) - a junction at 30% could get a 20% voltage malace or -0.2 voltage for say, the receiving lights, fabricators, engine, navigation etc). if the subs voltage is around 1.0, yet the power transfer junction was almost at 0%, at most this is half voltage and devices connected to it still functional, albeit in an awful state.

5.) Reactors either use fuel faster, or emit small amounts of damaging radiation leaks. Have a chance of a timed fire+explosion if the condition is under 10% yet the reactor is actively on based on availablefuel. I don't feel they should emit less power and simply use the fuel up faster with a chance of danger. shutting down a reactor to repair it is the most realistically safe thing you can do if its containment is almost breached.

6.) Engines have less output force (25%, 50% feels like it could be too excessive).

7.) Minimap and sonar already have the related effects of distorting and showing less data when damaged.

8.) Steering becomes slower and harder to adjust, perhaps reduces the max values by up to 20%.

9.) oxygen generator generates less total volume of oxygen (OR costs more power to operate).

10.) Vents (if they have deterioration say, via a mod) can output less oxygen volume.

11.) Railguns and controllers have no decay issues, but make ammo rack decay effect the reload timer, If you want decay on railguns, add it to the controller and have that decay effect accuracy and turn rate and bounce values. I know since they are linked this is rather easy to implement. well maintained ammo loaders means a good fire rate : >

12.) Light components should flicker and/or darken (If lights ever could get damaged IE mods, personally I wouldn't have them deteriorate nearly ever, maybe randomly break due to large ranging timers where the median is rare to break within an hour long round but let EMP strikes smash them).

13.) Docking ports transfer less power than asked for, yet still draw the full load.

14.) Doors could open slower when damaged or be prone to temporary jams (Add sparks or such to detail a door-jam).

15.) Electrical dischargers have no deterioration, but their range and capabilities could be impacted by damage to capacitors (battery components) they are linked to, capacitors usually output so much that less output wouldn't do much of anything anyways - but for dischargers they could have special checks.

The above makes losing a job a malace, not a ton of required work/a requirement to skill up to the skill requirement over an hour or a full game ender. A submarine will still move and operate minutes after a 20% condition fix on its engines for instance, albeit slower. a player may then optionally take the double time or something to hit the next stage of repair and improve engine capability, or focus on other tasks or things to do.

Essentially, don't implement "Bonuses" for constant fixing, but implement negatives for negligence over a long period, the stepped repairs means its harmless to leave things not repaired for a duration, so uneven timers is less of a direct issue and is more just occasional maintenance.

If a sub is run down all of these will stack on each other, such as reduced output from batteries, a reduction of voltage itself from the connected junction and then a damaged engine with reduced force, maintenance is important but not the strict focus.

- [ ] **Have those with insufficient skill only able to see repair icons at 50% condition or less when its performance is effected, unless the tool to fix it is equipped.**

The idea for this is most players won't see things that need maintenance unless their seeking it, while engineers will see engineer items and mechanics see mechanic items that could be given a quick fixer when passing by, everyone however will see a problem component and thus be able to report it as its damage sprites begin fading in, making it feel less of a concern for most and less obtrusive.

Perhaps if not holding the tools required for it, require your mouse to get close before seeing the icon as well, as if needing to look at and observe it. I know I would prefer this personally, but I imagine others would not. I would like a less obtrusive view myself.

***This may sound unnecessarily complicated but for the above, a new player only needs to understand two things, items need routine maintenance but can be left alone for a long period, and that condition below 50% will begin to impact performance of the submarine component, giving multiple chances to perform maitenance***

- [ ] **Implement scaling of repair timers to the size of the crew.**

In a 3 player game, repairs can be way too punishing, yet in a 10+ player game, there is too many people for there to be anything to even do or fix. the timers rate of decreased could be based on the count of players, perhaps even utilizing the submarines recommended player count or count of decay-able repairable devices vs player count, this should however include dead but spawned crew (unique crew) for the count but not spectators. if you count dead players, leaving the game and returning would simply reduce the repair requirements as I do not think the game hands back corpses.

I wouldn't scale the single player repairs to the crew size - the AI is more than happy to run about repairing after all.

- [ ] **Implement a server setting to enable, or disable repairable deterioration.**

For servers that want sub vs sub or traitor modes, they may want either more or less distraction - and I feel this should be an option to hosts. one can simply have the repairable timers not tick down if the setting is disabled, and also thus be able to enable this mid-round if wanted and retain water/emp/overload damages naturally without much of any code changes.

***Malfunction events and other in-gamely disaster - yes code for this already exists:***

- [ ] **Implement the Malfunction event and have it target a zone (like an EMP explosion) of devices simultaneously - Introduce an electric effect and sparks or a small explosion depending on the device (?) for said malfunction to be apparent.**

This can simply start by targeting a device inside a submarine at random and range selecting devices near that device or connected to it. they could also be given weights to be more likely to target certain devices, and ranges the malfunction can go up to (or max transferred devices to also malfunction from a random selection on the same network).

The current implementation of malfunction event is not like this, but i'm not sure I want 10 or even 5 things breaking in 3-10 different areas of a submarine to run left right and center for 5-10 minutes, I feel the above would be more suitable, as well as make it more obvious a malfunction is actively occurring from sound and visual.

**Comment by Regalis11** [comment]
Thanks for a well though-out suggestion once again (and to Ol Stinky as well)!

We're aware that the deterioration system still needs quite a lot of work, but I do believe the negative response to it is _partially_ because of the repair duration bug (27917ee) and due to players misinterpreting damage done by power issues as spontaneous deterioration. For instance, it should take 18 to 32 minutes for a junction box to break by itself, so these cases of boxes breaking constantly as you repair them should not happen as a result of the spontaneous deterioration. So I'm guessing either the system is bugged (which I'll do some more testing on shortly), or the problems are actually caused by the grid being overloaded. The automatic temperature control is generally quite slow to adjust the power output, so it's very easy to cause a chain reaction of breaking junction boxes if no-one is watching the reactor.

I do agree with most of your points and the ones Stinky made in his doc, but a few points stuck out:

>Have deterioration work in stages, so the first deterioration lowers it by 20% and initiates another timer, say in 20% stages), have the first repair available at the first step lost (80%)
Currently the devices gradually deteriorate to 0% during a span of about 16 minutes (or at least they should, if that's not the case the system must be bugged), and I don't quite see how changing it to discreet "steps" would affect or improve the gameplay.

>Modify the effectiveness of equipment based on its condition beyond a certain point (IE. below 50%, this should take multiple missed repairs).
This is already in the game, basically all deteriorating items become less effective when their condition decreases.

**Comment by NilanthAnimosus** [comment]
just a small note but this does not feel like im losing 1% condition every 10 seconds, but maybe I am...I didn't try timing if that really was 16 odd minutes.

At least, if i'm losing 0.10 / second with a max health of 100. then isn't that 100 / 0.1 for 1000 seconds? or alternatively to read it as its coded, 100 - (0.1 multiplied by time passed). is it just my imagination and bad math that it breaks faster than a good 16 minutes?

[Clip of deterioration on a junction with a 0.1 decay rate](https://i.gyazo.com/7775772e3b6e1defac3ee368d2771ec4.mp4)

csharp code:
```csharp
if (item.Condition > MinDeteriorationCondition)
{
     item.Condition -= DeteriorationSpeed * deltaTime;
}
```

**Comment by NilanthAnimosus** [comment]
Pumps, oxygen generators and engines appear to have this (More power + less output), though I feel deterioration shouldn't begin to have effects until below 50%, up to at most 50% of the effectiveness as you progress to 0% and then break. and not really cost more power as much as to waste it by being less useful.

But many components (reactors, fabricators, deconstructors, power containers, doors, turrets, electrical dischargers and such) do not have anything to them.

Finally, power transfer has the worst - where I will quote the code:
```csharp
//items in a bad condition are more sensitive to overvoltage
                float maxOverVoltage = MathHelper.Lerp(Math.Min(OverloadVoltage, 1.0f), OverloadVoltage, item.Condition / item.MaxCondition);
```

I'd prefer small, component-specific changes without modifying the power draws or resistance to voltage spike (Their already damaged and draining health after all), and for power transfer to well, be less effective at transferring power. the receiving items only need check the condition of the component it receives its power from after all (its direct link).

This may be why we think deterioration is bad, it converts those 2.3 overload voltages to 1.0f, while the reactor fluctuates power and almost anything breaking or not needing power will easilly drive loads straight up. 3700 going to 2500 is a 50% voltage spike and thats very common. as parts deteriorate they increase in power draw, while junctions decrease in voltage resistance, when something 'pop's or the reactor fluctuates and the sub is moving then not moving the junctions take heavy damage likely pretty fast.

The reason I thought of the steps was the first two chances to repair the item have zero effects on how it performs, while providing the icon to fix it, the same can really be achieved by a single timer, but giving a new, fresh timer each time means shorter timers between each repair chance.

Essentially, a random time between 2 minutes to 16 minutes, or 5 times 2 minutes to 4 minutes. just an example, I can repair objects more inline with each other and be more certain it'l last for XY time.

I can fix and reset its timer there and then, so if the room appears fixed, I know at worst its going to decrease by 20-40%. not 100% if I don't go back to it within a set time, depending what the decay rate of the item is. it seems odd to say it now given the time for a lot of things to break is a gradual change over 10 minutes but players get irritated just seeing the need to repair every 30 seconds they leave the room, and this does change how effective it is.

I genuinely feel if everyone can fix a junction in 10 seconds for 16.6 minutes, with an added time of 2-16 minutes, that engineers are not even needed, my above idea was to make them still very useful, but not strictly required to finish. if you dont fully repair it just continues decaying for now. the steps let it decay to a certain value where your sure you again have a timer.

Players spend more attention realizing there are things to repair than how long it actually takes to break really.

**Comment by Regalis11** [comment]
We're moving feature requests from the issue tracker to the Barotrauma forums. From now on only actual issues (bugs, crashes, design flaws, oversights) should be reported here. We're doing this because the issue tracker has started to become extremely cluttered - it's getting harder for us to find the more critical issues when they're mixed in with hundreds of ideas and feature requests. And even though we are closing the existing feature requests, we are still planning on implementing many of them further down the line. 

So from now on, please post your feature requests on the [Collaboration section on the Barotrauma forums](https://undertowgames.com/forum/viewforum.php?f=19&sid=3ef7ed701ca460dc3503e2ab4e3cf7c9).
